### PR TITLE
Improve Subprocess Usage and Output Parsing

### PR DESCRIPTION
**Pull request by rmartin16**
## Parsing `check_output` Return Values
In certain uses of `check_output`, issues can arise when processing the command output (as opposed to the command itself failing). Because the errors occur after `check_output` returns, the logging in #704 will not help users understand these errors.

These changes implement a new wrapper for `check_output` to invoke a supplied parser function on the command output. If the parsing fails, then the command, command output, and failure reason will be logged.

The calls to `simctl` on macOS were updated to use this new functionality.

### Implementation Overview
- Created `parse_output` as a wrapper for `check_output`
- `parse_output` takes an additional argument `parser`
  - `parser` is a function that takes a string as its first argument and returns the parsed data
  - If `parser` encounters an error, it should raise `ParserError` so `parse_output` can catch it and initiate error logging 
  - Error logging includes the command, its output, and a failure reason from `parser`
  - For parsing errors, `parse_output` will raise `CommandOutputParseError` for the calling code.
- A general parser for JSON was implemented for arbitrary callers.

## Defaulting `text=True` for all Subprocess calls
I also updated the usage of subprocess to include `text=True` by default. I did this for two reasons:
1) All calls to subprocess where the output was explicitly referenced, `universal_newlines` was being explicitly passed as True. (I did find an exception to this in `platforms/macOS/__init__.py`, but using `text=True` avoids the need for it to manually decode the bytes.)
2) I realized that the logging from #704 will disregard line breaks in output if `text` (or `universal_newlines`) is not set to `True`. This is because `str(b"...").splitlines()` doesn't work at all like we want.

If we don't want this new `text` default applied to all `subprocess` calls, I can revert this and add in specific checks in the `subprocess` logging to `decode` any output returned as bytes.

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by rmartin16** [comment]
@freakboy3742, when you get a chance, I would appreciate your feedback on the high level decisions I've made in this PR.
 - Any tweaks to the check_output parsing?
 - Other `check_output` uses that should implement this?
   - After reviewing them, I didn't really see near as much value as with the `simctl` calls.
   - I initially get create a parser function for `AndroidSDK.devices()`.....but it felt a little overdone since it didn't ever seem likely to meaningfully fail there.
 - Yay or nay to `text=true`?
   - One concern I have is if `errors` or `encoding` need appropriate defaults as well.

Additionally, while I don't own any apple products, I was surprisingly able to finally get macOS running in a virtualbox vm. That'll hopefully let me get better familiarity with briefcase on macOS.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
tl;dr: this is looking really good.

The `text` change makes me a *little* twitchy, because it means the Briefcase subprocess shim isn't a 100% reproduction of the underlying API - but then, we're also adding environment wrappers, Docker wrappers, and now parsing functionality as well. On balance, I think it's the right thing to do (on the basis that every existing usage needs `text=True`, and you've provided an escape hatch on the off chance it *isn't* needed).

Your instinct/question around defaulting stderr->stdout is also a good one. On the one hand, almost every usage of check_output does this redirection; on the other hand, stray stderr output in parsed content may be one of the causes of parsing failure (and *not* redirecting would likely protect us from parsing errors). Ironically, the specific problem that caused this PR (the error that occurs once after an Xcode update) is impossible to reproduce, so I can't answer whether avoiding stderr redirection would actually be a fix for the problem. My inclination is to leave the redirection as is for now, and revisit it if/when we see consistent errors that we can attribute to stderr redirection.

I'm not sure I have any immediate suggestions of other parsers that should be added - The JSON parser is the only one that is a format that is re-used in multiple locations. The other places where check_output is being parsed tends to be "single purpose" - i.e., the parser would be used exactly once (for extracting a version number for a specific tool, or a list from the output of a specific tool). There might be a *minor* testing benefit from factoring out the parser from the invocation; but honestly, I'm not sure it's worth the code churn. 

As for the meat of the PR - the actual `parse_output()` function: Fundamentally, this looks great (and almost exactly what I would have expected to see). The only problem/issue I can see is around output related to error handling.

As currently implemented, invalid JSON output will:
* cause a JSON parsing error,
* raises a ParserError,
* is caught, causing `log.error` output,
* is raised as a CommandOutputParseError,
* is caught, and converted into a BriefcaseCommandError

The final error that is printed is a high-level "Unable to invoke simctl" or similar error. My concern is that the CommandOutputParseError is almost invisible; and a parsing error will manifest as a "can't run" command, rather than a "can't parse" error. The CommandOutputParseError should be caught independently to the CalledProcessError so that a different contextual error message can be output.

File: tests/integrations/subprocess/test_Subprocess__check_output.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def test_call_with_arg(mock_sub, capsys):
 
     mock_sub._subprocess.check_output.assert_called_with(
         ['hello', 'world'],
-        universal_newlines=True
```
[the remark this answers is not in the document]
Comment (rmartin16) [reply tests/integrations/subprocess/test_Subprocess__check_output.py:25]: Added these tests for `run`, `check_output`, and `Popen`.

**Comment by rmartin16** [comment]
Updated to catch `CommandOutputParseError` so a different contextual error is presented to the user. And built out coverage for these changes. I also updated the `Subprocess` logging to ensure command outputs are strings.

I'm still trying to get the iOS simulator to work though....will submit once that's done.

**Review by freakboy3742 — APPROVED** [review]
Looks great. I've made a couple of minor cleanups on the way to landing - a couple of tuples that should be lists, plus some genericization of tests. 

The only change that is really substantial is the change in parameter ordering in parse_output - this is to ensure that the function can be curried/partialed (which isn't likely to be a common use case, but changing the order makes it at least possible).

Thanks for another great contribution!
